src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/EETypeVTableNode.cs

```diff
@@ -0,0 +1,115 @@
+            _type = type;
+        }
+
+        protected override void OnMarked(NodeFactory context)
+        {
+            if (context.CompilationModuleGroup.ShouldProduceFullType(_type))
+            {
+                ShouldBuildFullVTable = true;
+            }
+
+            if (ShouldBuildFullVTable)
+            {
+                BuildFullVTable(context);
+            }
+        }
+
+        public List<MethodDesc> Slots
+        {
+            get
+            {
+                return _slots;
+            }
+        }
+
+        public void AddEntry(NodeFactory context, MethodDesc virtualMethod)
+        {
+            if (ShouldBuildFullVTable)
+                return;
+
+            Debug.Assert(virtualMethod.IsVirtual);
+            
+            if (!_slots.Contains(virtualMethod))
+            {
+                _slots.Add(virtualMethod);
+            }
+        }
+        
+        private void BuildFullVTable(NodeFactory context)
+        {
+            MetadataType mdType = _type.GetClosestMetadataType();
+            
+            // On-demand VTable generation may have assigned slots already
+            _slots.Clear();
+            
+            foreach (var method in mdType.GetMethods())
+            {
+                if (!method.IsVirtual)
+                    continue;
+
+                context.VirtualMethodUse(method);
+                _slots.Add(method);
+            }
+        }
+
+        public override bool StaticDependenciesAreComputed
+        {
+            get
+            {
+                return true;
+            }
+        }
+        
+        public override string GetName()
+        {
+            return "__vtable_" + NodeFactory.NameMangler.GetMangledTypeName(_type);
+        }
+
+        public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFactory context)
+        {
+            List<DependencyListEntry> dependencies = new List<DependencyListEntry>();
+            if (_type.HasBaseType)
+            {
+                dependencies.Add(new DependencyListEntry(context.VTable(_type.BaseType), "Base type VTable"));
+            }
+
+            if (ShouldBuildFullVTable)
+            {
+                // When building a full VTable (such as for a type from a library), add dependencies on
+                // the virtual methods so that methods overriding them will be marked in the graph
+                // and compiled.
+                MetadataType mdType = _type.GetClosestMetadataType();
+                foreach (MethodDesc decl in VirtualFunctionResolution.EnumAllVirtualSlots(mdType))
+                {
+                    dependencies.Add(new DependencyListEntry(context.VirtualMethodUse(decl), "VTable method dependency"));
+                }
+            }
+
+            return dependencies;
+        }
+
+        public override void EncodeData(ref ObjectDataBuilder dataBuilder, NodeFactory factory, bool relocsOnly)
+        {
+            throw new NotImplementedException();
```

Comment:
> What this really is at the moment is a sort of graph expanding helper. Perhaps we need an abstraction for that.

ObjectNode doesn't force you to define a symbol (it doesn't implement ISymbolNode). ShouldSkipEmittingObjectNode lets you skip it from object emission. ComputeNonRelocationBasedDependencies lets you define dependencies that don't come from GetData.

That, plus an implementation of GetData that returns an empty ObjectData should work IMHO.